common/src/main/java/net/whimxiqal/journey/search/function/EuclideanPlanarAverageCostFunction.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
+ * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package net.whimxiqal.journey.search.function;
+
+import net.whimxiqal.journey.Cell;
+import net.whimxiqal.journey.Journey;
+
+public class EuclideanPlanarAverageCostFunction implements CostFunction {
+  @Override
+  public CostFunctionType getType() {
+    return CostFunctionType.EUCLIDEAN_PLANAR_AVERAGE;
+  }
+
+  private final static double SQRT_TWO = Math.sqrt(2);
+  private final static double SQRT_THREE = Math.sqrt(3);
+
+  private final Cell destination;
+
+  public EuclideanPlanarAverageCostFunction(Cell destination) {
+    this.destination = destination;
+  }
+
+  @Override
+  public Double apply(Cell cell) {
+    double euclideanCost = cell.distanceTo(destination);
+
+    // Sort coordinates in order of size
+    int xDiff = Math.abs(destination.blockX() - cell.blockX());
+    int yDiff = Math.abs(destination.blockY() - cell.blockY());
+    int zDiff = Math.abs(destination.blockZ() - cell.blockZ());
+    int first, second, third;
+    boolean isYFirst = false;
+    if (yDiff > xDiff && yDiff > zDiff) {
+      first = yDiff;
+      isYFirst = true;
+      if (xDiff > zDiff) {
+        second = xDiff;
+        third = zDiff;
+      } else {
+        second = zDiff;
+        third = xDiff;
+      }
+    }
+    else if (xDiff > zDiff) {
+      first = xDiff;
+      if (yDiff > zDiff) {
+        second = yDiff;
+        third = zDiff;
+      } else {
+        second = zDiff;
+        third = yDiff;
+      }
+    }
+    else {
+      first = zDiff;
+      if (xDiff > yDiff) {
+        second = xDiff;
+        third = yDiff;
+      } else {
+        second = yDiff;
+        third = xDiff;
+      }
+    }
+
+    // Calculate distance to travel
+    double total = SQRT_THREE * third;
+    second -= third;
+    first -= third;
+
+    total += SQRT_TWO * second;
+    first -= second;
+
+    if (isYFirst) total += first * SQRT_TWO;
+    else total += first;
+
+    double planarCost = total;
+
+    // Average together Euclidean distance and Planar cost.
+    double euclideanExponent = 1.35;
+    double planarMulti = 1.0;
+
+    double avgCost = (Math.pow(euclideanCost, euclideanExponent) + (planarCost * planarMulti)) / 2.0;
+
+    // Try to avoid caves.
+    if (destination.blockY() > 50)
+      avgCost *= Math.max(1.0, Math.log(Math.min(destination.blockY(), 55.0) / Math.max(1.0, cell.blockY())) * 8);
+
+    return avgCost;
```

Comment:
This is interesting. What is the motivation behind this approach? Why does averaging two equations give a better result? Also, why do you want to avoid caves? If a block in a cave gives a closer approximation to the destination than a block above ground, why not evaluate the block in the cave as a better option?

Reply:
I average the results because it tends to result in a shorter cone, so while it's less accurate, it checks significantly less blocks to get up/around an obstacle.

As for avoiding caves, when testing out the algorithm from random points I found that a lot of the time it would wander into a cave, then get stuck there because it has to basically fill up the entire cave before determining that there's not a valid path through.

With the current implementation however, it's more that it avoids caves that go down that avoiding them altogether, and since caves that go all the way through probably don't go that far down it'll probably still find it's way through. I can label it better if you'd like.